### PR TITLE
Being able to add buttons to the right of the subject line in thread view

### PR DESCRIPTION
**Pull request by ruthhalteman**
Streak example is our streak share button. Next to gmail's print and open in new buttons.

Currently draft because I don't know how correct or robust this is. Will keep testing. So far the main issue I see is that it's not finding the container in the reading pane views. Not sure why, the subject line label (existing functionality) is fine and works in a very similar way.

Summary of additions:
- Thread view function `addSubjectButton` that takes in a button descriptor and makes a button using our Button view
- Mutation observer for cases when the subject button container is lazy loaded
- Fix for case when a `ButtonView` doesn't have a dropdown where the focus class doesn't get removed

[Relevant Streak box](https://www.streak.com/a/boxes/agxzfm1haWxmb29nYWVyNAsSDE9yZ2FuaXphdGlvbiIRb2lzbWFpbEBnbWFpbC5jb20MCxIEQ2FzZRiAgO7r38zdCQw)

Known issues:
- [x] Spacing between SDK buttons isn't good/doesn't match gmail buttons (when adding two or more)

### CODE REVIEW

File: src/platform-implementation-js/dom-driver/gmail/views/gmail-thread-view.ts [hunk 1]
```diff
@@ -625,6 +627,25 @@ class GmailThreadView {
...
   }
 
+  addSubjectButton(button: any) {
+    const subjectToolbarElement = this._findSubjectToolbarElement();
+    if (!subjectToolbarElement) {
+      throw new Error('Subject toolbar element not found');
+    }
+    // gmail has these spacer spans. Not sure if important or a good idea to use this way
+    const el = document.createElement('span');
+    el.setAttribute('data-unique-tt-id', 'ucc-2');
```
Comment (ruthhalteman) [inline src/platform-implementation-js/dom-driver/gmail/views/gmail-thread-view.ts:637]: oh yeah and it shouldn't be hardcoded to 2 if this is actually kept in here. should be consecutive like the gmail ones are

File: src/platform-implementation-js/views/conversations/thread-view.ts [hunk 2]
```diff
@@ -143,6 +143,10 @@ class ThreadView extends (EventEmitter as new () => TypedEventEmitter<ThreadView
   addLabel(): SimpleElementView {
     return get(memberMap, this).threadViewImplementation.addLabel();
   }
+
+  addSubjectButton(buttonDescriptor: Record<string, any>) {
```
Comment (wegry) [inline src/platform-implementation-js/views/conversations/thread-view.ts:147]: Could we use https://github.com/InboxSDK/InboxSDK/blob/4edf7ccd5e69417933291d69006d111f7bfc4bf7/src/inboxsdk.d.ts#L446 here?

```suggestion
  addSubjectButton(buttonDescriptor: Descriptor<ButtonDescriptor>) {
```

File: src/platform-implementation-js/dom-driver/gmail/views/gmail-thread-view.ts [hunk 3]
```diff
@@ -687,6 +708,12 @@ class GmailThreadView {
     return null;
   }
 
+  _findSubjectToolbarElement(): HTMLElement | null | undefined {
```
Comment (wegry) [inline src/platform-implementation-js/dom-driver/gmail/views/gmail-thread-view.ts:711]: Unless we want this method visible outside the class...

```suggestion
  #findSubjectToolbarElement(): HTMLElement | null {
```

File: src/platform-implementation-js/views/conversations/thread-view.ts (hunk 2 above)
Comment (wegry) [reply src/platform-implementation-js/views/conversations/thread-view.ts:147]: If we don't necessarily want an `Observable` as part of the API, an alternative would be 

```suggestion
  addSubjectButton(buttonDescriptor: ButtonDescriptor) {
```

File: src/platform-implementation-js/dom-driver/gmail/views/gmail-thread-view.ts [hunk 4]
```diff
@@ -625,6 +627,25 @@ class GmailThreadView {
     return view;
   }
 
+  addSubjectButton(button: any) {
```
Comment (wegry) [inline src/platform-implementation-js/dom-driver/gmail/views/gmail-thread-view.ts:630]: We might use either `ButtonDescriptor` or `Descriptor<ButtonDescriptor>` as the type here.

**Review by wegry — COMMENTED** [review]
Looks pretty good so far.

I did notice the subject button disappearing after 5s or so with the reading pane enabled...

See https://www.loom.com/share/0dc58ca527294d7598dffb813600ae6d

File: src/platform-implementation-js/dom-driver/gmail/views/gmail-thread-view.ts [hunk 5]
```diff
@@ -687,6 +737,12 @@ class GmailThreadView {
     return null;
   }
 
+  _findSubjectToolbarElement(): HTMLElement | null {
+    var toolbarContainerElements =
+      this._element.querySelectorAll<HTMLElement>('.bHJ');
+    return toolbarContainerElements[0];
```
Comment (wegry) [inline src/platform-implementation-js/dom-driver/gmail/views/gmail-thread-view.ts:743]: What's the rationale for querySelectorAll here?

Maybe we just want:

```suggestion
    const toolbarContainerElement =
       this._element.querySelector<HTMLElement>('.bHJ');
     return toolbarContainerElement;
```

**Review by wegry — CHANGES_REQUESTED** [review]
Alright, so running through this example extension with the reading pane enabled and not looks good.

Few things I'd like to see changed prior to pulling it in are (in order):

1. Use `SelectorError` for reporting's sake when an element isn't present and we expect it to be so.
1. Plumb the ButtonDescriptor all the way through from the public interface (removing some remaining `any`s).
2. Not adding new non public methods to gmail-thread-view).

**Review by wegry — APPROVED** [review]

File: src/platform-implementation-js/views/conversations/thread-view.ts [hunk 6]
```diff
@@ -164,6 +164,10 @@ class ThreadView extends (EventEmitter as new () => TypedEventEmitter<ThreadView
         });
       });
   }
+
```
Comment (wegry) [inline src/platform-implementation-js/views/conversations/thread-view.ts:167]:
```suggestion
  /**
   * @internal
   */
```

**Pull request merged**